Lecture 3

sazro19 opened the pull request.

sazro19 committed:
Task 3.1

sazro19 committed:
Task 3.2

sazro19 committed:
Task 3.3

sazro19 committed:
Task 3.4

sazro19 committed:
Added Tests

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
+        lock.readLock().lock();
```
session attributes can be changed in this method, it's not really a read operation.

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,72 @@
+        try {
+            Product product = productDao.getProduct(productId);
+            CartItem cartItem = new CartItem(product, quantity);
+            if (cart.getItems().contains(cartItem)) {
```
here you rely on equals implementation, but actually you need only product id to find item

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,72 @@
+            Product product = productDao.getProduct(productId);
+            CartItem cartItem = new CartItem(product, quantity);
+            if (cart.getItems().contains(cartItem)) {
+                int index = cart.getItems().indexOf(cartItem);
```
not a fan of indexOf...

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,72 @@
+            }
+            if (isStockNotAvailable(product, quantity)) {
+                throw new OutOfStockException(product, quantity, product.getStock());
+            }
```
[ opt ] you can move this check up for optimization purposes

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/viewed/RecentlyViewedContainer.java:
```diff
@@ -0,0 +1,42 @@
+            }
+            return;
+        }
+        if (threeLastProducts.size() == 3) {
```
make 3 constant

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -1,34 +1,94 @@
-        Long productId = Long.valueOf(request.getPathInfo().substring(1));
+        Long productId = parseProductId(request);
+        RecentlyViewedContainer container = recentlyViewedService.getContainer(request);
+        recentlyViewedService.add(container, productId);
```
I have two methods that looks almost the same: container.addProduct and recentlyViewedService#add
Which one should I use?

Alexandra-ES replied:
Now it looks even more scary and error-prone

Alexandra-ES commented on src/main/java/com/es/phoneshop/web/ProductDetailsPageServlet.java:
```diff
@@ -1,34 +1,94 @@
+            NumberFormat format = NumberFormat.getInstance(request.getLocale());
+            quantity = format.parse(quantityString).intValue();
+        } catch (ParseException e) {
+            request.setAttribute("error", "Not a number");
```
make "error" constant

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Override
+    public Cart getCart(HttpServletRequest request) {
```
in this method you need only session, not entire request

Alexandra-ES commented on src/main/webapp/WEB-INF/pages/productList.jsp:
```diff
@@ -46,4 +46,24 @@
     </c:forEach>
   </table>
+  <p>
+    <b>Recently viewed</b>
```
recently viewed section can be in separate tag

sazro19 committed:
- added constants;
- method getCart and getContainer require only session;
- added tag with recently viewed products table;
- product displayed in recently viewed section after it opened on PDP;

sazro19 committed:
- Changed add method in DefaultCartService;
- addProduct in RecentlyViewedContainer is protected;

sazro19 committed:
- Added SessionLock;

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/cart/DefaultCartService.java:
```diff
@@ -46,20 +47,22 @@ public void add(Cart cart, Long productId, int quantity) throws OutOfStockExcept
-                int oldQuantity = cart.getItems().get(index).getQuantity();
+            Optional<CartItem> containedCartItem = cart.getItems().stream()
+                    .filter(cartItem -> cartItem.getProduct().getId().equals(product.getId()))
+                    .findFirst();
```
use findFirst() only when you specifically want the first element from a sequence, otherwise use findAny()

Alexandra-ES commented on src/main/webapp/WEB-INF/pages/product.jsp:
```diff
@@ -64,21 +64,5 @@
-      </th>
-    </c:forEach>
-  </table>
+  <tags:recentlyViewed recentlyViewedList="${viewed.threeLastProducts}"/>
```
in JSP you only need threeLastProducts, no need to pass entire container

Alexandra-ES commented on src/main/java/com/es/phoneshop/model/product/viewed/RecentlyViewedContainer.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    public Deque<Product> getThreeLastProducts() {
+        return threeLastProducts;
```
Make defensive copy and return it

sazro19 committed:
- changed doGet() PDP servlet, added getThreeRecentlyViewedProducts method;
- changed to findAny() in DefaultCartService;
- changed "viewed" attribute in product.jsp and productList.jsp;
